File: LDE.c
#include <stdio.h>
#include <stdlib.h>
#include <conio.h>
#include <locale.h>

typedef struct No{
    char info;
    struct No *eloA, *eloP;
} LDE;

void inserirLDE( LDE **comeco, LDE **fim, char letra ){
    LDE *novo, *ant, *prox;

    novo = malloc( sizeof(LDE) );
    novo->info = letra;
    novo->eloA = NULL;
    novo->eloP = NULL;

    if( (*comeco) == NULL ){ // Caso 1
        (*comeco) = novo;
        (*fim) = novo;
        return;
    }
    if( letra < (*comeco)->info ){ // Caso 2
        novo->eloP = (*comeco);
        (*comeco)->eloA = novo;
        (*comeco) = novo;
        return;
    }
    if( letra > (*fim)->info ){ // Caso 3
        (*fim)->eloP = novo;
        novo->eloA = (*fim);
        (*fim) = novo;
        return;
    }
    // Caso 4
    ant = (*comeco);
    while( ant != NULL ){
        prox = ant->eloP;
        if( letra > ant->info && letra < prox->info ){
            ant->eloP = novo;
            novo->eloP = prox;
            prox->eloA = novo;
            novo->eloA = ant;
            return;
        }
        ant = ant->eloP;
    }
}

int retirarLDE( LDE **comeco, LDE **fim, char letra ){
    LDE *aux;

    aux = (*comeco);
    while( aux != NULL ){

        if( aux->info == letra ){

            if( aux == (*comeco) ){
                if( aux->eloP == NULL ){ // Caso 1 - único elemento
                    (*comeco) = NULL;
                    (*fim) = NULL;
                }
                else{ // Caso 2 - início da lista
                    (*comeco) = aux->eloP;
                    (*comeco)->eloA = NULL;
                }
            }
            else{
                if( aux == (*fim) ){ // Caso 3 - final da lista
                    (*fim) = aux->eloA;
                    (*fim)->eloP = NULL;
                }
                else{ // Caso 4 - meio da lista
                    LDE *ant, *prox;
                    ant = aux->eloA;
                    prox = aux->eloP;
                    ant->eloP = prox;
                    prox->eloA = ant;
                }
            }
            free( aux );
            return 1;
        }
        aux = aux->eloP;
    }
    return 0;
}

void mostrarLDE_C(char frase[], LDE *comeco ){
    LDE *aux;

    printf( "\n%s: ",frase );
    aux = comeco;
    while( aux != NULL ){
        printf( "%c", aux->info );
        aux = aux->eloP;
        if( aux != NULL ) printf( " -> " );
    }
    printf( "\n" );
}

void mostrarLDE_D(char frase[], LDE *fim ){
    LDE *aux;

    printf( "\n%s: ",frase );
    aux = fim;
    while( aux != NULL ){
        printf( "%c", aux->info );
        aux = aux->eloA;
        if( aux != NULL ) printf( " -> " );
    }
    printf( "\n" );
}

int main(){
    setlocale(LC_ALL, "Portuguese");
    LDE *comeco=NULL, *fim=NULL, *aux;
    char letra, op;

    do{
        printf( "\nOpção (I-inserir//R-retirar//C-mostrar cres.//D-mostrar desc.//F-fim): " );
        op = getche();
        switch( op ){
            case 'I': case 'i':
                printf( "\nLetra para inserção: " );
                letra = getche();
                inserirLDE( &comeco, &fim, letra );
                break;
            case 'R': case 'r':
                printf( "\nLetra para retirada: " );
                letra = getche();
                if( retirarLDE( &comeco, &fim, letra ) )
                    printf( "\nRetirada realizada com sucesso!" );
                else
                    printf( "\nErro: Letra não presente na lista!" );
                break;
            case 'C': case 'c':
                mostrarLDE_C( "LDE Crescente", comeco );
                break;
            case 'D': case 'd':
                mostrarLDE_D( "LDE Decrescente", fim );
                break;
        }
    }while( op != 'F' && op != 'f' );
    printf( "\n" );
}

File: ListaLDE/src/listalde/LDE.java
package listalde;

public class LDE 
{
    public No comeco;
    public No fim;
    
    public LDE()
    {
        comeco = null;
        fim = null;
    } 

    public boolean vazia()
    {
        return (comeco==null);
    }

    public void inserir( int exp, double coeficiente  )
    {
        No novo =  new No(exp, coeficiente);
        
        if( comeco == null )// caso 1
        { 
            comeco = novo;
            fim = novo;
            return;
        }
        if( exp > comeco.exp )// Caso 2
        { 
            comeco.eloA = novo;
            novo.eloP = comeco;
            comeco = novo;
            return;
        }
        if( exp < fim.exp )// Caso 3
        { 
            fim.eloP = novo;
            novo.eloA = fim;
            fim = novo;
            return;
        }
        // Caso 4 se já existe fazer o que deseja, subtrair coeficienter ou multipicar
        No ant = comeco;
        while( ant != null ){
            No prox = ant.eloP;
            if( exp == ant.exp){
                ant.coef += coeficiente;
                return;
            }
            if( exp < ant.exp && exp > prox.exp )
            { 
                ant.eloP = novo;
                novo.eloP = prox;
                prox.eloA = novo;
                novo.eloA = ant;  
                return;
            }
            ant = ant.eloP;
        }
    }

    public LDE multiplicar(LDE f1, LDE f2)
    {
        No l1 = f1.comeco;
        No l2 = f2.comeco;
        LDE f3 = new LDE();
        int exp;
        double coef;
        while(l1 != null){
            l2 = f2.comeco;
            while(l2 != null){
                exp = l1.exp + l2.exp;
                coef = l1.coef * l2.coef;
                f3.inserir(exp, coef);
                l2 = l2.eloP;
            }
            l1 = l1.eloP;
        }
        
        return f3;
    }

    public LDE somarLista(LDE f1, LDE f2)
    {
        LDE f3 = new LDE();
        No l1 = f1.comeco;
        No l2 = f2.comeco;
                
        while(l1 != null)
        {
            f3.inserir(l1.exp, l1.coef);
            l1 = l1.eloP;
        }
        while (l2 != null)
        {
            f3.inserir(l2.exp, l2.coef);
            l2 = l2.eloP;
        }
        
        return f3;
    }
    
    
    public LDE subtrairLista(LDE f1, LDE f2)
    {
        LDE f3 = new LDE();
        No l1 = f1.comeco;
        No l2 = f2.comeco;
        
        while(l1 != null || l2 != null)
        {
            if(l1 !=null && l2 != null){
                if(l1.exp == l2.exp)
                    {
                        f3.inserir(l1.exp, l1.coef - l2.coef);
                        l1 = l1.eloP;
                        l2 = l2.eloP;

                    }
            }
            else 
            {
                if(l1.exp > l2.exp)
                {
                    f3.inserir(l1.exp, l1.coef);
                    l1 = l1.eloP;
                }
                else if(l2.exp > l1.exp)
                {
                    f3.inserir(l2.exp, l2.coef);
                    l2 = l2.eloP;
                }
            }
        }
        
        return f3;
    }
    
    
    public void mostrar(String frase){
        No aux;
        System.out.print( frase );
        aux = comeco;
        while( aux != null )
        {
            if(aux.eloP == null && aux.exp >= 0) {
                if (aux.exp != 0)System.out.print(aux.coef + "X^" + aux.exp);
                else System.out.print(aux.coef + " ");
                }
            else if(aux.exp > 0) System.out.print( aux.coef + "X^" + aux.exp + " + ");
            else if(aux.exp == 0) System.out.println(aux.coef +" + ");
            else if(aux.exp < 0) System.out.print(" + " + aux.coef + "X^" + aux.exp );
            aux = aux.eloP;
        }
        System.out.println();
    }
    
}